Add cmdstanr compatibility

n-kall opened the pull request:
The latests cmdstanr (not yet released, but on github) has functionality for `log_prob` and `unconstrain_pars` methods. So I adapted the `stanfit` method to one for `CmdStanFit`.

Commit:
minor fix

Commit:
add moment matching for cmdstanr models

Commit:
add tests for cmdstanr models

Commit:
rename functions

Commit:
add posterior as dependency, add cmdstanr as suggests

Commit:
add constrain_all_pars functions

topipa commented:
Thanks for opening this PR! I added you as a collaborator now in this repository. Let me know if you want me to review this before merging :slightly_smiling_face:

Commit:
add new test

Commit:
run tests in parallel

Commit:
rename functions *_pars -> *_draws to be clearer

Commit:
documentation

n-kall committed:
Merge branch 'topipa:main' into main

Commit:
add new tests for cmdstanr

Commit:
Merge branch 'main' of github.com:n-kall/iwmm into main

Commit:
add r cmd check workflow

Commit:
fix workflow files

Commit:
fix workflow file

Commit:
yamllint fixes

Commit:
rename yaml

Commit:
indentation fix

Commit:
fix cmdstan workflow

Commit:
add github token to action

Commit:
tweak workflow

Commit:
typofix

Commit:
remove old rstan and cmdstanr tests

Commit:
default constrain draws to true

Commit:
update tests

Commit:
rename files

Commit:
clean up rstan and cmdstanr code

Commit:
fix tests

Commit:
minor fix, formatting

Commit:
documentation

Commit:
minor tweak to formatting in vignette

Commit:
documentation

Commit:
add helpers file

Commit:
fix dependencies in workflow

Commit:
delint vignette

Commit:
update workflow

Commit:
update docs

Commit:
delint functions

Commit:
delint tests

Commit:
update suggests

Commit:
minor fixes to vignette

Commit:
remove workflows

Commit:
readd lint workflow

Commit:
readd test coverage workflow

Commit:
fix dependencies

Commit:
fix coverage workflow

Commit:
update workflow

Commit:
update workflow

Commit:
typo

Commit:
update rbuildignore

Commit:
minor fix to tests

Commit:
remove parallelisation in vignette to get checks to pass

Commit:
update contributors

Commit:
export constrain_draws method

topipa commented on tests/testthat/test-moment-match-stanfit-analytical.R:
```diff
@@ -0,0 +1,175 @@
+  # ratio = jointlikelihood
+  joint_log_lik <- function(draws, fit, ...) {
+
+    cdraws <- constrain_draws_stanfit(fit, draws)
```
I get a warning about this, should it be just constrain_draws?

n-kall replied:
Fixed

topipa commented on tests/testthat/test-moment-match-cmdstan-analytical.R:
```diff
@@ -0,0 +1,168 @@
+  # ratio = jointlikelihood
+  joint_log_lik <- function(draws, fit, ...) {
+
+    cdraws <- constrain_draws_CmdStanFit(fit, draws)
```
I get a warning about this, should it be just constrain_draws?

n-kall replied:
Fixed

topipa commented on R/cmdstanr_functions.R:
```diff
@@ -0,0 +1,109 @@
+#' Generic importance weighted moment matching algorithm for `stanfit` objects.
+#' See additional arguments from `moment_match.matrix`
```
These should say `CmdStanFit`, right?

n-kall replied:
fixed

topipa commented on R/moment_match.R:
```diff
@@ -114,7 +114,8 @@ moment_match.matrix <- function(x,
     k <- lw_psis$diagnostics$pareto_k
-    if (is.infinite(k)) {
+    if (any(is.infinite(k))) {
+      print(exp(lw))
```
Is this print intentional?

n-kall replied:
Not intentional, it should be fixed now

topipa reviewed (approved):
Overall this looks good to me :+1: Thanks for improving the structure and adding the workflows about linting and coverage etc.

I left a few small comments. The old tests in `test-moment-match-stanfit.R` can also be made to work with relatively small changes. I can clean them up and add them after this is merged, for now your analytical tests will do just fine.

n-kall commented:
I haven't done much with this as I am waiting for the cmdstanr unconstrain_pars etc. interface to be finalised. I'll then get back to dealing with your comments. :)

rantahar committed:
Enable caching built R packages (#1)

* Cache CmdStanR

* Fix indentation

* Fix indentation in Cache step

* Fix path name

* Set cache version to 1 in setup-r-dependencies.

Otherwise the existing cache is ignored

* Change key to match naming convention.

Also trigger to see if anything is created.

* Add a simple test

Making sure I have some idea about how caching works...

* Try adding shell

* cat -> echo and try a working example

* Fix path variable

* Delete cache test

* Separate build steps into another workflow.
Also trigger these from R-CMD-check, since it needs the results. These should be cached at that point.

* Add local_actions to the local action

* The local action must be called action.yaml

and be in it's own folder

* Rename

* Follow a different structure example

* Update rest of the yaml strucutre

* Don't try cancelling in the local action

* Use variables in the call

* Try using cached cmdstanr in other workflows

* Use the build action correctly in R-CMD-check

* Fix cmdstanr path

* Skip cmdstanr build step if cache found

* Removed caches, retry

* Try without the if and see what happens

* Add the if statement to prevent a warning

* Move cmdstanr installation to a separate action

* Remove the additional environment change

* Try action in action

* Add separate manual build workflows for each original workflow

* Replace missing action call with content

* Descriptive names for cache actions

* Identifiers for the build workflows

* Fix test-coverage workflows

* Remove extra parameters

Commit:
update (un)constrain functions for cmdstanr

Commit:
add abind dependency

Commit:
namespace fixes

Commit:
lint vignette

Commit:
lint functions

Commit:
linting

rantahar committed:
Remove the auto_write rstan option

to avoid bug with R 4.2 on windows.

rantahar committed:
Add worflow_dispatch to workflows so that I can test run

rantahar committed:
Add missing :

rantahar committed:
Reorder to avoid misreading the name

rantahar committed:
Check without cmdstanr.

This shuld be sufficient for CRAN.

rantahar committed:
Check without extra packages

rantahar committed:
Disable the cmdstanr test directly

rantahar committed:
Disable cmdstanR isntall checks

Trying a solution from discourse.mc-stan.org

rantahar committed:
Try without fix=TRUE

rantahar committed:
Back to normal, try removing caches

rantahar committed:
Disable caching cmdstanR setup

this does not depend on the version, so it can cause issues.

rantahar committed:
Conditional test (#4)

* Move cmdstanr to enhancements and add repo

This enables running the tests without the package installed

* Add a guard to skip the cdmstanr tests when not installed

* Also run the tests before installing cmdstanr

* Fix new linting warnings

* Remove additional repository

Misread this, stan-dev is not an R repository.

Commit:
fix merge conflict

Commit:
Merge branch 'AaltoRSE-windows_error' into main

Commit:
adjust linting

Commit:
fix linter settings

Commit:
linting

Commit:
formatting

Commit:
remove print statement

Commit:
update documentation

n-kall commented:
@topipa I think this is now in a good state to merge. Then we can think about a CRAN release. Would be great if you can take a look at the new tests and see if you think they make sense

Commit:
shorten line length

n-kall commented:
> Would be great if you can take a look at the new tests and see if you think they make sense

I just realised you already checked the tests last review, somehow I missed that. So I don't think anything major changed, just kept up to date with new methods in cmdstanr

n-kall commented:
Currently the tests are failing on Windows, but perhaps we can deal with that after the merge

topipa reviewed (approved):
@n-kall still looks good. Feel free to merge this.
